### PR TITLE
Utility functions to draw shapes that visualise tensors and vectors

### PR DESCRIPTION
**Pull request by ceriottm**
Added utility functions to visualize positive-definite tensors (polarizability, thermal ellipsoids, ...) as well as vectors (forces, dipoles, ....). The latter required changing (switching off) the centering of the shapes. I think this is better, because you can always center (and add a utility function for that) a custom shape, but you can't draw an off-center shape if vertices get centered automatically. 

Also fixed `chemiscope.show` so it takes a `shapes` argument.

### CODE REVIEW

File: python/chemiscope/structures/_shapes.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            ]
+        )
+    # Compute the eigenvalues and eigenvectors
+    eigenvalues, eigenvectors = np.linalg.eigh(matrix)
+
+    if force_positive:
+        eigenvalues = np.abs(eigenvalues)
+
+    # The lengths of the principal axes are the square roots of the eigenvalues
+    old_settings = np.seterr(invalid="raise")
```
Comment (rosecers) [inline python/chemiscope/structures/_shapes.py:155]: Why are you bypassing this?

File: python/jupyter/src/widget.ts [hunk 2]
```diff
@@ -48,6 +48,7 @@ class ChemiscopeBaseView extends DOMWidgetView {
             },
             this
         );
+        this._updatePythonSettings();
```
Comment (rosecers) [inline python/jupyter/src/widget.ts:51]: Why this change?

File: python/chemiscope/structures/_shapes.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    points -= points.mean(axis=0)
+    new_shape["vertices"] = points.tolist()
+
+    return new_shape
+
+
+def _oriented_circle(radius, nvec, n_points=20):
+    """Generates a circle in 3D centered at the origin and oriented according to the given normal"""
+    # makes sure the normal is unit
+    nvec = nvec / np.linalg.norm(nvec)
```
[the remark this answers is not in the document]
Comment (ceriottm) [reply python/chemiscope/structures/_shapes.py:26]: Won't do - it changes nvec in place which is problematic. I will rather edit the code to have a different variable name so there's no risk of confusion.

File: python/chemiscope/structures/_shapes.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+    # generate n_points in the plane defined by nvec, centered at vec
+    angles = np.linspace(0, 2 * np.pi, n_points, endpoint=False)
+    circle_points = np.outer(np.cos(angles), u) + np.outer(np.sin(angles), v)
+    circle_points = radius * circle_points
+
+    return circle_points
+
+
+def arrow_from_vector(
```
[the remark this answers is not in the document]
Comment (ceriottm) [reply python/chemiscope/structures/_shapes.py:47]: shading is OK, if not perfect. can we provide normals?

File: python/chemiscope/structures/_shapes.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        if m00 < -m11:
+            t = 1 - m00 - m11 + m22
+            q = np.array((m20 + m02, m12 + m21, t, m01 - m10))
+        else:
+            t = 1 + m00 + m11 + m22
+            q = np.array((m12 - m21, m20 - m02, m01 - m10, t))
+
+    q /= np.sqrt(q @ q)  # makes sure the quaternion is normalized
+
+    return q
```
[the remark this answers is not in the document]
Comment (ceriottm) [reply python/chemiscope/structures/_shapes.py:123]: OK. Didn't want to add a dependency on scipy just for this, but I see we have already one for the convex hull, so OK.

File: python/chemiscope/structures/_shapes.py (hunk 1 above)
Comment (ceriottm) [reply python/chemiscope/structures/_shapes.py:155]: I'm not bypassing it. By default it's a warning, and I'm turning it to an error.

File: python/jupyter/src/widget.ts (hunk 2 above)
Comment (ceriottm) [reply python/jupyter/src/widget.ts:51]: as per commit comment - otherwise on the python side the settings are uninitialized unless accessed from the notebook, so when you show and then chemiscope.save, you'll get empty settings otherwise.

Comment (ceriottm) [reply python/jupyter/src/widget.ts]: ok, my bad. I dug deeper and this was a red herring. problem was elsewhere, it's now fixed and this line has gone.

Comment (ceriottm) [reply python/jupyter/src/widget.ts]: code review done right, thank you!

File: python/chemiscope/structures/_shapes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ceriottm) [reply python/chemiscope/structures/_shapes.py:46]: I think arrows look decent - perhaps we can take care of normals in a PR in which we revamp the shape definition format.

**Comment by ceriottm** [comment]
@rosecers  I think I addressed all your comments. Nod if you're OK with merging now so we can move on to the next PR!

File: python/chemiscope/structures/_shapes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Luthaf) [reply python/chemiscope/structures/_shapes.py:46]: > shading is OK, if not perfect. can we provide normals?

in modern graphics API, normals are implicitly provided by the order of the indices/simplices. They are assumed to be in clockwise or counter-clockwise order, and the screw axis of the rotation is taken as the normal. The only thing I'm not sure about is if 3dmol is using clockwise or counter-clockwise ...

**Comment by Luthaf** [comment]
Regarding arrows: should we have these as predefined shape type inside chemiscope? This will save a lot of space by not requiring to store all the vertices/simplices in the file, but instead generate them on the JS side like we do for ellispoids.

File: python/chemiscope/structures/_ase.py [hunk 6]
```diff
@@ -304,6 +305,66 @@ def _is_convertible_to_property(value):
...
+    :param frames: list of ASE Atoms objects
+    :param key: name of the ASE atom property. Should contain
+       three components corresponding to x,y,z
+    """
+
+    vectors = []
+
+    for f in frames:
+        if key not in f.arrays:
+            raise IndexError(f"Key {key} not found in Atoms frame")
```
Comment (Luthaf) [inline python/chemiscope/structures/_ase.py:325]:
```suggestion
            raise IndexError(f"Key {key} not found in `Atoms.arrays`")
```

this will avoid some confusion if people thing we are looking for stuff in `info`

File: python/chemiscope/structures/_shapes.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        ax = np.sqrt(eigenvalues[0])
+        ay = np.sqrt(eigenvalues[1])
+        az = np.sqrt(eigenvalues[2])
+    except FloatingPointError:
+        print(
+            "Non-positive definite tensor found with eigenvalues ",
+            eigenvalues,
+            ".\nIf this is acceptable, set `force_positive=True` to take the absolute values.",
+        )
+        raise
```
Comment (Luthaf) [inline python/chemiscope/structures/_shapes.py:138]:
```suggestion
    except FloatingPointError as e:
        raise ValueError(
            f"Non-positive definite tensor found with eigenvalues {eigenvalues}.\n"
            "If this is acceptable, set `force_positive=True` to take the absolute values.",
        ) from e
```

let's not use print as a way to handle/propagate errors

File: python/chemiscope/structures/_shapes.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    # makes sure the rotation is proper
+    if np.linalg.det(eigenvectors) < 0:
+        eigenvectors[:, 0] *= -1
+
+    # Form the rotation matrix from eigenvectors
+    rotation = eigenvectors.T
+
+    # converts to quaternion
+    from scipy.spatial.transform import Rotation
```
Comment (rosecers) [inline python/chemiscope/structures/_shapes.py:149]: Move this to the top of the file

**Review by rosecers — COMMENTED** [review]
I like @Luthaf's suggestion of an arrow as, essentially, a primitive type!

**Comment by ceriottm** [comment]
> Regarding arrows: should we have these as predefined shape type inside chemiscope? This will save a lot of space by not requiring to store all the vertices/simplices in the file, but instead generate them on the JS side like we do for ellispoids.

I agree, but I would like to do this as part of a separate PR in which we address the input format too. In the spirit of doing small incremental PRs.

File: python/chemiscope/structures/_shapes.py (hunk 8 above)
Comment (ceriottm) [reply python/chemiscope/structures/_shapes.py:147]: This is imported in place as for the part that computes simplices. Let's avoid a global dependency, particularly if we're going to make the arrows a JS-side primitive.

Comment (ceriottm) [reply python/chemiscope/structures/_shapes.py]: I added an exception check

**Review by Luthaf — APPROVED** [review]

**Comment by Luthaf** [comment]
> I agree, but I would like to do this as part of a separate PR in which we address the input format too. In the spirit of doing small incremental PRs.

This is tracked in #291!
